### PR TITLE
ntfsck: check and repair $I30 index entries

### PR DESCRIPTION
**Pull request by jaysim**
Check $I30 index entries (directory entries) with MFT entry's $FILE_NAME attributes.
 - check parent MFT reference
 - check allocation size / real size
 - check reparse point tag

If they are different, copy MFT entry's $FILE_NAME attribute to Index Entry's $FILE_NAME key.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

### CODE REVIEW

File: ntfsprogs/ntfsck.c [hunk 1]
```diff
@@ -453,8 +559,101 @@ static int ntfsck_check_name_attr_index_entry(ntfs_inode *ni, INDEX_ENTRY *ie,
...
+				(unsigned long long)ni->mft_no, filename);
+			need_fix = 1;
+			goto fix_index;
+		}
+	}
+
+	/* set NI_FileNameDirty in ni->state to sync
+	 * $FILE_NAME attrib when ntfs_inode_close() is called */
+fix_index:
+	if (need_fix) {
```
Comment (namjaejeon) [inline ntfsprogs/ntfsck.c:638]: We don't need to ask if it could be fixed using ntfsck_ask_repair() here?

File: ntfsprogs/ntfsck.c [hunk 2]
```diff
@@ -390,30 +485,37 @@ struct ntfsls_dirent {
...
 	MFT_REF mref = le64_to_cpu(ie->indexed_file);
-	int err;
-	char *s;
+	FILE_NAME_ATTR *fn,
+		       *first_fn = NULL;
+	FILE_NAME_ATTR *ie_fn = (FILE_NAME_ATTR *)&ie->key;
+	ATTR_RECORD *attr;
+	char *filename;
+	int ret = -EIO;
+	int need_fix = FALSE;
```
Comment (namjaejeon) [inline ntfsprogs/ntfsck.c:504]: It can be declared like this : BOOL need_fix = FALSE;

File: ntfsprogs/ntfsck.c [hunk 3]
```diff
@@ -453,8 +559,101 @@ static int ntfsck_check_name_attr_index_entry(ntfs_inode *ni, INDEX_ENTRY *ie,
...
+			ntfs_attr_name_free(&filename);
+			goto remove_index;
+
+		} else if (idx_pdir_seq != mft_pdir_seq) {
+			check_failed("Parent MFT's sequence No. is differnt "
+					"(IDX/$FN:%u MFT/$FN:%u) "
+					"on inode(%llu, %s)\n",
+					idx_pdir_seq, mft_pdir_seq,
+					(unsigned long long)ni->mft_no, filename);
+			need_fix = 1;
```
Comment (namjaejeon) [inline ntfsprogs/ntfsck.c:593]: need_fix = TRUE;

File: ntfsprogs/ntfsck.c [hunk 4]
```diff
@@ -378,6 +378,101 @@ static void ntfsck_verify_mft_record(ntfs_volume *vol, s64 mft_num)
...
+	ni = actx->ntfs_ino;
+
+	std_info = (STANDARD_INFORMATION *)((u8 *)actx->attr +
+			le16_to_cpu(actx->attr->value_offset));
+	si_ctime = std_info->creation_time;
+	si_mtime = std_info->last_data_change_time;
+	si_mtime_mft = std_info->last_mft_change_time;
+	si_atime = std_info->last_access_time;
+
+	int jump = 0;
```
Comment (namjaejeon) [inline ntfsprogs/ntfsck.c:411]: You need to declare variable at the top of function on C lang, It will cause build warning.

Comment (namjaejeon) [reply ntfsprogs/ntfsck.c:411]: and we can use BOOL type for 0 or 1(TRUE, FALSE)

File: ntfsprogs/ntfsck.c [hunk 5]
```diff
@@ -378,6 +378,101 @@ static void ntfsck_verify_mft_record(ntfs_volume *vol, s64 mft_num)
...
+	printf("std_info ctime:%llx, mtime:%llx, mftime:%llx, atime:%llx\n",
+			si_ctime, si_mtime, si_mtime_mft, si_atime);
+	printf("mft_fn ctime:%llx, mtime:%llx, mftime:%llx, atime:%llx\n",
+			mft_fn->creation_time, mft_fn->last_data_change_time,
+			mft_fn->last_mft_change_time, mft_fn->last_access_time);
+	printf("idx_fn ctime:%llx, mtime:%llx, mftime:%llx, atime:%llx\n",
+			idx_fn->creation_time, idx_fn->last_data_change_time,
+			idx_fn->last_mft_change_time, idx_fn->last_access_time);
+	printf("======== END =======================\n");
+	return;
```
Comment (namjaejeon) [inline ntfsprogs/ntfsck.c:472]: we don't need return here.

File: ntfsprogs/ntfsck.c [hunk 6]
```diff
@@ -378,6 +378,101 @@ static void ntfsck_verify_mft_record(ntfs_volume *vol, s64 mft_num)
...
+		jump = 1;
+	}
+
+	if (idx_fn->reparse_point_tag != mft_fn->reparse_point_tag) {
+		printf("different reparse_point IDX/$FN:%x, MFT/$FN:%x\n",
+				idx_fn->reparse_point_tag,
+				mft_fn->reparse_point_tag);
+		jump = 1;
+	}
+	if (jump)
```
Comment (namjaejeon) [inline ntfsprogs/ntfsck.c:453]: if (!jump)
      return;

And we can remove print_time goto statement.

File: ntfsprogs/ntfsck.c [hunk 7]
```diff
@@ -378,6 +378,101 @@ static void ntfsck_verify_mft_record(ntfs_volume *vol, s64 mft_num)
...
+	}
+
+	if (si_mtime != idx_fn->last_data_change_time ||
+			si_mtime_mft != idx_fn->last_mft_change_time) {
+		printf("STD TIME != IDX/$FN\n");
+		jump = 1;
+	}
+
+	if (idx_fn->parent_directory != mft_fn->parent_directory) {
+		printf("different parent_directory IDX/$FN, MFT/$FN\n");
```
Comment (namjaejeon) [inline ntfsprogs/ntfsck.c:431]: Probably we can use ntfs_log_debug() ?

File: ntfsprogs/ntfsck.c [hunk 8]
```diff
@@ -390,30 +485,37 @@ struct ntfsls_dirent {
 NTFS_LIST_HEAD(ntfs_dirs_list);
 
 
-static int ntfsck_check_name_attr_index_entry(ntfs_inode *ni, INDEX_ENTRY *ie,
+/* check $FILE_NAME attribute in directory index and same one in MFT entry
```
Comment (namjaejeon) [inline ntfsprogs/ntfsck.c:488]: Looks good to add the block comment!
But Block comments use a trailing /* on a separate line.
```
 /*
  * check $FILE_NAME attribute in directory index and same one in MFT entry
  * Blablah...
  * /
```

File: ntfsprogs/ntfsck.c [hunk 9]
```diff
@@ -453,8 +559,101 @@ static int ntfsck_check_name_attr_index_entry(ntfs_inode *ni, INDEX_ENTRY *ie,
...
+	 * $FILE_NAME attrib when ntfs_inode_close() is called */
+fix_index:
+	if (need_fix) {
+		NInoFileNameSetDirty(ni);
+		NInoSetDirty(ni);
+		ntfs_inode_update_times(ictx->ni, NTFS_UPDATE_CTIME);
+
+		/* CHECK: copy all MFT/$FN field to IDX/$FN */
+		memcpy(ie_fn, fn, sizeof(FILE_NAME_ATTR) - sizeof(ntfschar));
+		ret = ntfsck_write_index_entry(ictx);
```
Comment (namjaejeon) [inline ntfsprogs/ntfsck.c:645]: if ntfsck_write_index_entry() return error, we don't need to decrement error count(errors--).

File: ntfsprogs/ntfsck.c (hunk 1 above)
Comment (jaysim) [reply ntfsprogs/ntfsck.c:638]: we need it! I'll add that check routine.

**Comment by jaysim** [comment]
I'll check all comments, and re-request it.

**Comment by jaysim** [comment]
I checked all comments and modified code.
And code for allocation size fix logs was checked in binary format through comparison with ntfsprogs and windows chkdsk execution if it's value is right.

**Comment by jaysim** [comment]
I modified some condition about parent directories value of $FN between (IDX, MFT entry and parent inode)

**Comment by namjaejeon** [comment]
Applied, Thanks for your work!

**Pull request closed without merging**